src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1284 @@
+            public override BoundNode? VisitRelationalPattern(BoundRelationalPattern node)
+            {
+                TryPushOperand(NegateIfNeeded(node));
+                return null;
+            }
+
+            // Given a pattern that expects `pattern.InputType` as input type, produce
+            // one that expects `inputType.
+            //
+            // For example, when splitting a pattern `{ Prop: A and B }` into `{ Prop: A } and/or { Prop: B }`
+            // the original `B` pattern expects an input type that is the narrowed type from `A`
+            // but the rewritten `B` pattern expects an input type that is the type of `Prop`.
+            private static BoundPattern WithInputTypeCheckIfNeeded(BoundPattern pattern, TypeSymbol inputType)
+            {
+                if (pattern.InputType.Equals(inputType, TypeCompareKind.AllIgnoreOptions))
+                {
+                    return pattern;
+                }
+
+                if (pattern is BoundTypePattern typePattern1)
+                {
+                    return typePattern1.Update(typePattern1.DeclaredType, typePattern1.IsExplicitNotNullTest, inputType, typePattern1.NarrowedType);
+                }
+
+                if (pattern is BoundRecursivePattern recursivePattern)
+                {
+                    return recursivePattern.Update(
+                        recursivePattern.DeclaredType ??
+                            new BoundTypeExpression(recursivePattern.Syntax, aliasOpt: null, recursivePattern.InputType.StrippedType()),
+                        recursivePattern.DeconstructMethod, recursivePattern.Deconstruction,
+                        recursivePattern.Properties, recursivePattern.IsExplicitNotNullTest,
+                        recursivePattern.Variable, recursivePattern.VariableAccess,
+                        inputType, recursivePattern.NarrowedType);
+                }
+
+                if (pattern is BoundDiscardPattern discardPattern)
+                {
+                    return discardPattern.Update(inputType, inputType);
+                }
+
+                if (pattern is BoundNegatedPattern negatedPattern)
+                {
+                    return negatedPattern.Update(
+                        WithInputTypeCheckIfNeeded(negatedPattern.Negated, inputType),
+                        inputType, inputType);
+                }
+
+                Debug.Assert(pattern is not BoundBinaryPattern);
+
+                // Produce `PatternInputType and pattern` given a new input type
+
+                BoundPattern typePattern = new BoundTypePattern(pattern.Syntax,
+                    new BoundTypeExpression(pattern.Syntax, aliasOpt: null, pattern.InputType),
+                    isExplicitNotNullTest: false, inputType, narrowedType: pattern.InputType).MakeCompilerGenerated();
+
+                var result = new BoundBinaryPattern(pattern.Syntax, disjunction: false, left: typePattern, right: pattern, inputType, pattern.NarrowedType);
```

Comment:
>BoundBinaryPattern [](http://example.com/codeflow?start=33&length=18)

For constant and relational patterns would it be valid to simply adjust InputType in the node without creating an and?